### PR TITLE
Ensure that Quarkus can mock final methods of beans created by producers

### LINKED ISSUE #26733: Mocked bean created via method is not injected in another bean
### Describe the bug

Beans which are created via `@Produces` api are not injected in others beans fields. It works fine if I annotate bean with `@ApplicationScope` instead of using `@Produces`.

### Expected behavior

Mocked bean in passed to other beans.

### Actual behavior

Real bean in passed to other beans.

### How to Reproduce?

example repo: https://github.com/vadim-hleif/test-quarkus-mocking 

### Output of `uname -a` or `ver`

Darwin Vadims-MacBook-Pro.local 20.6.0 Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:42 PDT 2021; root:xnu-7195.141.8~1/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk 17.0.3 2022-04-19 LTS OpenJDK Runtime Environment Zulu17.34+19-CA (build 17.0.3+7-LTS) OpenJDK 64-Bit Server VM Zulu17.34+19-CA (build 17.0.3+7-LTS, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

7.4.2

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #26733

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java [hunk 1]
```diff
@@ -283,6 +284,20 @@ void implementDelegate(ClassCreator clientProxy, ProviderType providerType, Fiel
...
             BytecodeCreator falseBranch = creator.ifNull(mock).falseBranch();
+
+            ResultHandle mockGetClass = falseBranch
+                    .invokeVirtualMethod(MethodDescriptor.ofMethod(Object.class, "getClass", Class.class), mock);
+            ResultHandle mockClassLoader = falseBranch.invokeVirtualMethod(
+                    MethodDescriptor.ofMethod(Class.class, "getClassLoader", ClassLoader.class), mockGetClass);
+            ResultHandle mockClassLoaderToString = falseBranch
+                    .invokeVirtualMethod(MethodDescriptor.ofMethod(Object.class, "toString", String.class), mockClassLoader);
+            ResultHandle out = falseBranch.readStaticField(FieldDescriptor.of(System.class, "out", PrintStream.class));
+            falseBranch.invokeVirtualMethod(MethodDescriptor.ofMethod(PrintStream.class, "println", void.class, String.class),
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java:295]: `System.out.println()`?

Comment (geoand) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java:295]: Ah darn. Debug statements left over :)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java [hunk 2]
```diff
@@ -352,6 +362,22 @@ Collection<MethodInfo> getDelegatingMethods(BeanInfo bean, Consumer<BytecodeTran
...
     }
 
+    private void addDelegatesAndTrasformIfNecessary(Consumer<BytecodeTransformer> bytecodeTransformerConsumer,
+            boolean transformUnproxyableClasses,
+            Map<Methods.MethodKey, MethodInfo> methods, IndexView index,
+            Map<String, Set<Methods.NameAndDescriptor>> methodsFromWhichToRemoveFinal,
+            ClassInfo fieldClass) {
+        Methods.addDelegatingMethods(index, fieldClass, methods, methodsFromWhichToRemoveFinal,
+                transformUnproxyableClasses);
+        if (!methodsFromWhichToRemoveFinal.isEmpty()) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java:372]: I wonder if we should only trigger transformations for normal scoped beans?

I.e. check the scope of the bean...

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java (hunk 1 above)
Comment (geoand) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java]: Fixed

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java (hunk 2 above)
Comment (geoand) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java:357]: Why do you think we should do that?

Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java:357]: Forget about it, we're talking about client proxies... so normal scoped beans only. My brain needs a restart.

File: extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/RequestScopedFinalMethodsTest.java [hunk 3]
```diff
@@ -26,7 +26,8 @@ public class RequestScopedFinalMethodsTest {
     @RegisterExtension
     public static QuarkusUnitTest container = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
-                    .addClasses(RequestScopedBean.class));
+                    .addClasses(RequestScopedBean.class, OtherRequestScopedBean.class, OtherRequestScopeBeanProducer.class,
+                            StringProducer.class, StringObserver.class));
```
Comment (mkouba) [inline extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/RequestScopedFinalMethodsTest.java:30]: Why do you add the `StringProducer.class` and `StringObserver.class` to the test? Is it used somewhere?

Comment (geoand) [reply extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/RequestScopedFinalMethodsTest.java:30]: Good point, they do seem useless... removing

**Review by mkouba — APPROVED** [review]
Looks good!

**Pull request merged**